scanner/fetcher.go

```diff
@@ -0,0 +1,193 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package scanner
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"sync"
+
+	ct "github.com/google/certificate-transparency-go"
+	"github.com/google/certificate-transparency-go/client"
+)
+
+// FetcherOptions holds configuration options for the Fetcher.
+type FetcherOptions struct {
+	// Number of entries to request in one batch from the Log.
+	BatchSize int
+
+	// Number of concurrent fetcher workers to run.
+	ParallelFetch int
+
+	// [StartIndex, EndIndex) is a log entry range to fetch. If EndIndex == 0,
+	// then it gets reassigned to sth.TreeSize.
+	StartIndex int64
+	EndIndex   int64
+
+	// Don't print any status messages to default logger.
+	Quiet bool
+}
+
+// DefaultFetcherOptions returns new FetcherOptions with sensible defaults.
+func DefaultFetcherOptions() *FetcherOptions {
+	return &FetcherOptions{
+		BatchSize:     1000,
+		ParallelFetch: 1,
+		StartIndex:    0,
+		EndIndex:      0,
+		Quiet:         false,
+	}
+}
+
+// Fetcher is a tool that fetches entries from a CT Log.
+type Fetcher struct {
+	// Client used to talk to the CT log instance.
+	client *client.LogClient
+	// Configuration options for this Fetcher instance.
+	opts *FetcherOptions
+
+	// Current STH of the Log this Fetcher sends queries to.
+	sth *ct.SignedTreeHead
+
+	// TODO(pavelkalinnikov): Consider log.Logger instead.
+	Log func(msg string)
+}
+
+// EntryBatch represents a contiguous range of entries of the Log.
+type EntryBatch struct {
+	start   int64          // LeafIndex of the first entry in the range.
+	entries []ct.LeafEntry // Entries of the range.
+}
+
+// fetchRange represents a range of certs to fetch from a CT log.
+type fetchRange struct {
+	start int64 // inclusive
+	end   int64 // inclusive
+}
+
+// NewFetcher creates a Fetcher instance using client to talk to the log,
+// taking configuration options from opts.
+func NewFetcher(client *client.LogClient, opts *FetcherOptions) *Fetcher {
+	fetcher := &Fetcher{client: client, opts: opts}
+	if opts.Quiet {
+		fetcher.Log = func(msg string) {}
+	} else {
+		fetcher.Log = func(msg string) { log.Print(msg) }
+	}
+	return fetcher
+}
+
+// Prepare caches the latest Log's STH in the Fetcher and returns it. It also
+// adjusts the entry range to fit the size of the tree.
+func (f *Fetcher) Prepare(ctx context.Context) (*ct.SignedTreeHead, error) {
+	sth, err := f.client.GetSTH(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetSTH() failed: %v", err)
+	}
+	f.Log(fmt.Sprintf("Got STH with %d certs", sth.TreeSize))
+	if f.opts.EndIndex == 0 || f.opts.EndIndex > int64(sth.TreeSize) {
+		f.opts.EndIndex = int64(sth.TreeSize)
+	}
+	f.sth = sth
+	return sth, nil
+}
+
+// Run performs fetching of the Log. Blocks until scanning is complete or
+// context is cancelled. For each successfully fetched batch, runs the out
```

Comment:
`out` is an odd name now this is a callback; `process` or `fn` ?

Reply:
Renamed to `fn` for brevity.